#! /bin/sh /usr/share/dpatch/dpatch-run
## allow-ext3-journal-move.dpatch by Phillip Susi <psusi@cfl.rr.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: This patch allows defrag to move the ext3 journal

@DPATCH@
diff -urNad defrag-0.73pjm1~/defrag.c defrag-0.73pjm1/defrag.c
--- defrag-0.73pjm1~/defrag.c	2006-11-16 21:42:11.922981000 -0500
+++ defrag-0.73pjm1/defrag.c	2006-11-16 21:42:12.066990000 -0500
@@ -52,7 +52,7 @@
 int readonly = 0;
 int changed = 0;
 int blocks_until_sync = 0;
-Block bad_block_inode = 0, user_bad_inode = 0;
+Block bad_block_inode = 0, user_bad_inode = 0, journal_inode;
 
 #ifndef FS_IS_ext2
 Block next_block_to_fill = 0;
@@ -440,7 +440,8 @@
 
 	for (i=1; i < FIRST_USER_INODE; i++) {        
 	    if ((i == EXT2_ROOT_INO) ||  /* Allow optimization of root */
-		(i == bad_block_inode))  /* Done with them already */
+		(i == bad_block_inode) ||  /* Done with them already */
+		(i == journal_inode))
 		continue;
 	    if (!inode_in_use (i))
 		die ("Reserved inode is on the free list.");
@@ -542,7 +543,7 @@
 		printf ("DEBUG: scan_used_inodes()\n");
 	if (verbose)
 		stat_line ("Scanning inode zones...");
-	for (i=1; i<=INODES; i++) {
+	for (i=FIRST_USER_INODE; i<=INODES; i++) {
 	  	if (inode_in_use(i) && (i != bad_block_inode))
 			optimise_inode(i, 1);
                                 /* Update at least 50 times during scan */
@@ -694,6 +695,8 @@
 	inode_order_map[used_inodes++] = ROOT_INO;
 	if (bad_block_inode)
 		inode_order_map[used_inodes++] = bad_block_inode;
+	if( journal_inode )
+	  inode_order_map[used_inodes++] = journal_inode;
 
 	for (i = FIRST_USER_INODE; i <= INODES; i++)
 	{
@@ -707,6 +710,8 @@
 	   blocks next */
 	if (bad_block_inode)
 		inode_priority_map[bad_block_inode] = 126;
+	if( journal_inode )
+	  inode_priority_map[journal_inode] = 126;
 
 	/* And sort... */
 	qsort (inode_order_map, used_inodes,
diff -urNad defrag-0.73pjm1~/ext2.c defrag-0.73pjm1/ext2.c
--- defrag-0.73pjm1~/ext2.c	2006-11-16 21:42:11.726968750 -0500
+++ defrag-0.73pjm1/ext2.c	2006-11-16 21:42:12.066990000 -0500
@@ -49,6 +49,7 @@
 #define bm_mark_zone(x) (setbit(zone_map,(x)-FIRSTZONE),changed=1)
 #define bm_unmark_zone(x) (clrbit(zone_map,(x)-FIRSTZONE),changed=1) 
 
+extern Block journal_inode;
 
 /** @postcondition E1: Super.s_inodes_per_group % 8 == 0.
     @postcondition E2: Super.s_blocks_per_group % 8 == 0.
@@ -118,6 +119,8 @@
 					| EXT2_FEATURE_INCOMPAT_FILETYPE)))
       die( "filesystem has unsupported features");
 
+    if(Super.s_feature_compat & EXT3_FEATURE_COMPAT_HAS_JOURNAL)
+      journal_inode = 8;
     block_size = EXT2_MIN_BLOCK_SIZE << Super.s_log_block_size;
 
     groups = (Super.s_blocks_count - Super.s_first_data_block + 
